sdk/nodejs/cmd/pulumi-language-nodejs/main.go

```diff
@@ -1222,18 +1222,46 @@ func (host *nodeLanguageHost) GetProgramDependencies(
 	// dependencies, we need to also parse "package.json" and intersect it with
 	// reported dependencies.
 	var err error
-	yarnFile := filepath.Join(req.Info.ProgramDirectory, "yarn.lock")
-	npmFile := filepath.Join(req.Info.ProgramDirectory, "package-lock.json")
-	packageFile := filepath.Join(req.Info.ProgramDirectory, "package.json")
+	packagePathCheck := func(path string) bool {
+		info, err := os.Stat(path)
+		if err != nil || info.IsDir() {
+			return false
+		}
+		name := info.Name()
+		return name == "yarn.lock" || name == "package-lock.json"
+	}
+	packagePath, err := fsutil.WalkUp(req.Info.ProgramDirectory, packagePathCheck, nil)
+	// We special case permission errors to cause ErrProjectNotFound to return from this function. This is so
+	// users can run pulumi with unreadable root directories.
+	var perr *fs.PathError
+	if errors.As(err, &perr) {
```

Comment:
Would `errors.Is(err, fs.ErrPermission)` not work here? If not, it might be worth adding a comment to explain why, since this feels unintuitive to me -- I'd (perhaps wrongly) expect `PathError` to wrap errors in a manner that `As`/`Is` could work with.

Reply:
I think it should. `Is` should unwrap each layer of the error chain looking for it's target.

Reply:
good point! tbh I copied this code from another spot in the codebase that uses fsutil.WalkUp() that I assumed was put there through hard experience (written by... Fraser? :)) https://github.com/pulumi/pulumi/blob/b85a92cd5c8222815d2e942c6fcb2f99f50e3f45/sdk/go/common/workspace/paths.go#L135

Reply:
Ha, yeh that's probably slightly wrong as well. It was a year ago, but I don't recall any reason why I did As then Is.